### PR TITLE
MetaClient Distributed Semaphore API and Permit object

### PR DESCRIPTION
**Pull request by Marcosrico**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

https://github.com/apache/helix/issues/2237

### Description

Distributed Semaphore API. The implementation will come as a separate pull request.
The permit object acts as a ticket or ownership of a semaphore.

### Tests

- [x] The following tests are written for this issue:

n/a

- The following is the result of the "mvn test" command on the appropriate module:

(If CI test fails due to known issue, please specify the issue and test PR locally. Then copy & paste the result of "mvn test" to here.)

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+import org.apache.helix.metaclient.api.MetaClientInterface;
+import org.apache.helix.metaclient.datamodel.DataRecord;
+import org.apache.helix.metaclient.factories.MetaClientConfig;
+import java.util.Collection;
+import java.util.concurrent.TimeUnit;
+
+public class DistributedSemaphore {
+
+  /**
+   * Create a distributed semaphore client and creates a semaphore at the indicated path with the given capacity.
```
Comment (xyuanlu) [inline meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java:32]: Thanks for making the change! nit: add param description as well.

File: meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+   * Create a distributed semaphore client. Connects to an existing semaphore at given path.
+   */
+  public DistributedSemaphore(MetaClientConfig config, String path) {
+    throw new NotImplementedException("Not implemented yet.");
+  }
+
+  /**
+   * Connect to an existing distributed semaphore client and semaphore at the indicated path.
+   */
+  public DistributedSemaphore(MetaClientInterface<DataRecord> client, String path) {
```
Comment (xyuanlu) [inline meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java:48]: We should have a constructor that also takes an existing client for producer as well?

File: meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  /**
+   * Get the semaphore data record
+   * @return semaphore data record
+   */
+  private DataRecord getSemaphore() {
+    throw new NotImplementedException("Not implemented yet.");
+  }
+
+  /**
+   * Return a permit. If the permit is already returned, it will be ignored and throw a log info.
```
Comment (xyuanlu) [inline meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java:99]: throw an exception...?

Comment (Marcosrico) [reply meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java:99]: Mind my weird phrasing but i meant to say we can just log that the permit has already been returned. Since the end result is the same for success and failure, I don't think throwing an exception is necessary.

File: meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java (hunk 2 above)
Comment (Marcosrico) [reply meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java:48]: I see...in this case it would make sense to have a create and retrieve semaphore methods separate to that of the constructor.

File: meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Marcosrico) [reply meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java:32]: Good question. We have considered that possibility. However, since a client is only mapped to one semaphore, it makes sense that the constructor creates it. This way the user just needs to call one method to create both the client and the semaphore.
However, it seems like we want to create and abstraction to these two methods, just out of convenience instead of having multiple constructors that might be confusing. In the future if we would like to make it so that a client can connect to multiple semaphores, it would be easier to change as well.
So in conclusion i'll change it to what you suggested. I think that makes more sense in the long run. Thanks for the comment!

[the remark this answers is not in the document]
Comment (xyuanlu) [reply meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java]: Thanks for the detailed comment! 
I am not sure if in the future we would like to have one client connect to multiple semaphore as we already defined the `path` in constructor, witch map to one semaphore object. Since we wont be changing the API param in the future, then probably the access pattern stays the same.

[the remark this answers is not in the document]
Comment (Marcosrico) [reply meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java]: I'm not quite sure if I follow. I updated the pull request with new changes where neither constructor takes in a `path` as argument. When creating a semaphore, the user can either create a client or use their own to create a semaphore. If a semaphore already exists they can call retrieveSemaphore with the specified path to get the semaphore. You are correct in the sense that it is just a one client to one semaphore mapping. However, in the future if we were to do a one to many mapping, the changes implementation wise wouldn't be as complicated (by having these two separate create and retrieve methods). I hope this makes sense!

**Review by xyuanlu — APPROVED** [review]

**Comment by Marcosrico** [comment]
PR approved by @xyuanlu 
Commit message: MetaClient Distributed Semaphore API and Permit object
